test/integration/cdi/example-driver/controller/controller.go

```diff
@@ -0,0 +1,749 @@
+	// allocations. If we find that any of the claims cannot be allocated
+	// for the selected node, we don't need to try for the others either
+	// and shouldn't, because those allocations might have to be undone to
+	// pick a better node. If we don't need to allocate now, then we'll
+	// simply report back the gather information.
+	if len(podScheduling.Spec.PotentialNodes) > 0 {
+		if err := ctrl.driver.UnsuitableNodes(ctx, pod, claims, podScheduling.Spec.PotentialNodes); err != nil {
+			return fmt.Errorf("checking potential nodes: %v", err)
+		}
+	}
+	selectedNode := podScheduling.Spec.SelectedNode
+	logger.V(5).Info("pending pod claims", "claims", claims, "selectedNode", selectedNode)
+	if selectedNode != "" {
+		unsuitable := false
+		for _, delayed := range claims {
+			if hasString(delayed.UnsuitableNodes, selectedNode) {
+				unsuitable = true
+				break
+			}
+		}
+
+		if unsuitable {
+			logger.V(2).Info("skipping allocation for unsuitable selected node", "node", selectedNode)
+		} else {
+			logger.V(2).Info("allocation for selected node", "node", selectedNode)
+			selectedUser := &corev1.ResourceClaimUserReference{
+				Version:  "v1",
+				Resource: "pods",
+				Name:     pod.Name,
+				UID:      pod.UID,
+			}
+			for _, delayed := range claims {
+				if err := ctrl.allocateClaim(ctx, delayed.Claim, delayed.ClaimParameters, delayed.Class, delayed.ClassParameters, selectedNode, selectedUser); err != nil {
+					return fmt.Errorf("allocation of pod claim %s failed: %v", delayed.PodClaimName, err)
+				}
+			}
+		}
+	}
+
+	// Now update unsuitable nodes. This is useful information for the scheduler even if
+	// we managed to allocate because we might have to undo that.
+	// TODO: replace with patching the array. We can do that without race conditions
+	// because each driver is responsible for its own entries.
+	modified := false
+	podScheduling = podScheduling.DeepCopy()
+	for _, delayed := range claims {
+		i := findClaim(podScheduling.Status.Claims, delayed.PodClaimName)
+		if i < 0 {
+			// Add new entry.
+			podScheduling.Status.Claims = append(podScheduling.Status.Claims, corev1.ResourceClaimSchedulingStatus{PodResourceClaimName: delayed.PodClaimName})
+			modified = true
+		} else if stringsDiffer(podScheduling.Status.Claims[i].UnsuitableNodes, delayed.UnsuitableNodes) {
+			// Update existing entry.
+			podScheduling.Status.Claims[i].UnsuitableNodes = delayed.UnsuitableNodes
+			modified = true
+		}
+	}
+	if modified {
+		if _, err := ctrl.kubeClient.CoreV1().PodSchedulings(podScheduling.Namespace).UpdateStatus(ctx, podScheduling, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("update unsuitable node status: %v", err)
+		}
+	}
+
+	// We must keep the object in our queue and keep updating the
+	// UnsuitableNodes fields.
+	return errPeriodic
+}
+
+type claimAllocations []*ClaimAllocation
+
+// MarshalLog replaces the pointers with the actual structs because
+// we care about the content, not the pointer values.
```

Comment:
This currently has no effect in klog because of https://github.com/kubernetes/klog/issues/343